util/blooms-db/src/db.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{io, fmt};
+use std::path::Path;
+
+use ethbloom;
+
+use file::{File, FileIterator};
+
+/// Bloom positions in database files.
+#[derive(Debug)]
+struct Positions {
+	top: u64,
+	mid: u64,
+	bot: u64
+}
+
+impl Positions {
+	fn from_index(index: u64) -> Self {
+		Positions {
+			top: index >> 8,
+			mid: index >> 4,
+			bot: index,
+		}
+	}
+}
+
+/// Blooms database.
+pub struct Database {
+	/// Top level bloom file
+	///
+	/// Every bloom represents 16 blooms on mid level
+	top: File,
+	/// Mid level bloom file
+	///
+	/// Every bloom represents 16 blooms on bot level
+	mid: File,
+	/// Bot level bloom file
+	///
+	/// Every bloom is an ethereum header bloom
+	bot: File,
+}
+
+impl Database {
+	/// Opens blooms database.
+	pub fn open<P>(path: P) -> io::Result<Database> where P: AsRef<Path> {
+		let path = path.as_ref();
+		let database = Database {
+			top: File::open(path.join("top.bdb"))?,
+			mid: File::open(path.join("mid.bdb"))?,
+			bot: File::open(path.join("bot.bdb"))?,
+		};
+
+		Ok(database)
+	}
+
+	/// Insert consecutive blooms into database starting with positon from.
+	pub fn insert_blooms<'a, I, B>(&mut self, from: u64, blooms: I) -> io::Result<()>
+	where ethbloom::BloomRef<'a>: From<B>, I: Iterator<Item = B> {
+		for (index, bloom) in (from..).into_iter().zip(blooms.map(Into::into)) {
+			let pos = Positions::from_index(index);
+
+			// constant forks make lead to increased ration of false positives in bloom filters
+			// since we do not rebuild top or mid level, but we should not be worried about that
+			// most of the time events at block n(a) occur also on block n(b) or n+1(b)
+			self.top.accrue_bloom::<ethbloom::BloomRef>(pos.top, bloom)?;
```

Comment:
Regarding possible bloomdb true negative corrupt situation, given that we only insert one bloom at a time, I think a simple solution maybe to create a metadata file, storing the current writing bloom position before attempting the insertion. Then, if we find any error case in below, next time when the program starts, we would only need to rewrite that particular bloom position to fix the corruption.

This doesn't solve the disk consistency issue. Otherwise we will need to `flush` within the for loop, which may be a bad performance penalty. So not sure whether we would want this.

Reply:
> Then, if we find any error case in below, next time when the program starts, we would only need to rewrite that particular bloom position to fix the corruption.

That's what I did initially. Then I thought about the problem again and realised that all of this does not matter. Existence of blooms is not consensus critical. If one of the writes fail (or partially fail), this particular block will be reimported on next launch of parity and the database will overwrite corrupted data